File: src/services/session_service.py
"""
Session Management Service
Handles PC usage sessions with real-time tracking and sync
"""

from typing import Dict, Optional
from datetime import datetime
import time
from PyQt6.QtCore import QTimer, QObject, pyqtSignal

from services.firebase_client import FirebaseClient
from utils.logger import get_logger

logger = get_logger(__name__)


class SessionService(QObject):
    """Manages user sessions with Firebase sync"""

    # Signals
    time_updated = pyqtSignal(int)  # Emits remaining seconds
    session_ended = pyqtSignal(str)  # Emits reason: 'expired', 'user', 'error'
    warning_5min = pyqtSignal()
    warning_1min = pyqtSignal()
    sync_failed = pyqtSignal(str)  # Emits error message
    sync_restored = pyqtSignal()

    def __init__(self, firebase_client: FirebaseClient, user_id: str):
        super().__init__()
        self.firebase = firebase_client
        self.user_id = user_id

        # Session state
        self.session_id: Optional[str] = None
        self.is_active = False
        self.remaining_time = 0  # seconds
        self.time_used = 0  # seconds
        self.start_time: Optional[datetime] = None

        # Sync state
        self.last_sync_time = 0
        self.sync_queue = []  # Queue updates when offline
        self.is_online = True
        self.consecutive_sync_failures = 0

        # Timers
        self.countdown_timer = QTimer()
        self.countdown_timer.timeout.connect(self._on_countdown_tick)

        self.sync_timer = QTimer()
        self.sync_timer.timeout.connect(self._sync_to_firebase)

        # Warning flags
        self.warned_5min = False
        self.warned_1min = False

        logger.info(f"Session service initialized for user: {user_id}")

    def start_session(self, initial_remaining_time: int) -> Dict:
        """
        Start a new session

        Args:
            initial_remaining_time: User's remaining time in seconds

        Returns:
            {'success': bool, 'session_id': str, 'error': str}
        """
        logger.info("Starting new session")

        if self.is_active:
            logger.warning("Session already active")
            return {
                'success': False,
                'error': 'Session already active'
            }

        if initial_remaining_time <= 0:
            logger.error("Cannot start session with 0 time")
            return {
                'success': False,
                'error': 'No time remaining'
            }

        # Create session in Firebase
        session_data = {
            'userId': self.user_id,
            'startTime': datetime.now().isoformat(),
            'lastHeartbeat': datetime.now().isoformat(),
            'isActive': True,
            'timeUsed': 0,
            'status': 'active'
        }

        # Generate session ID using Firebase push
        import requests
        response = requests.post(
            f"{self.firebase.database_url}/sessions.json",
            params={'auth': self.firebase.id_token},
            json=session_data
        )

        if response.status_code != 200:
            logger.error(f"Failed to create session: {response.text}")
            return {
                'success': False,
                'error': 'Failed to create session in database'
            }

        self.session_id = response.json()['name']
        logger.info(f"Session created: {self.session_id}")

        # Update user's active session
        self.firebase.db_update(f'users/{self.user_id}', {
            'activeSessionId': self.session_id
        })

        # Initialize local state
        self.remaining_time = initial_remaining_time
        self.time_used = 0
        self.start_time = datetime.now()
        self.is_active = True
        self.warned_5min = False
        self.warned_1min = False

        # Start timers
        self.countdown_timer.start(1000)  # Every second
        self.sync_timer.start(10000)  # Every 10 seconds

        logger.info(f"Session started with {initial_remaining_time}s")
        return {
            'success': True,
            'session_id': self.session_id
        }

    def end_session(self, reason: str = 'user') -> Dict:
        """
        End the current session

        Args:
            reason: 'user', 'expired', 'error'
        """
        if not self.is_active:
            return {'success': False, 'error': 'No active session'}

        logger.info(f"Ending session: {self.session_id} (reason: {reason})")

        # Stop timers
        self.countdown_timer.stop()
        self.sync_timer.stop()

        # Final sync
        self._final_sync(reason)

        # Update state
        self.is_active = False

        # Emit signal
        self.session_ended.emit(reason)

        logger.info(f"Session ended. Time used: {self.time_used}s")
        return {
            'success': True,
            'time_used': self.time_used,
            'remaining_time': self.remaining_time
        }

    def _on_countdown_tick(self):
        """Called every second to update countdown"""
        if not self.is_active:
            return

        # Decrement time
        self.remaining_time -= 1
        self.time_used += 1

        # Emit update
        self.time_updated.emit(self.remaining_time)

        # Check warnings
        if self.remaining_time == 300 and not self.warned_5min:
            logger.warning("5 minutes remaining")
            self.warned_5min = True
            self.warning_5min.emit()

        if self.remaining_time == 60 and not self.warned_1min:
            logger.warning("1 minute remaining")
            self.warned_1min = True
            self.warning_1min.emit()

        # Check expiration
        if self.remaining_time <= 0:
            logger.warning("Time expired")
            self.end_session('expired')

    def _sync_to_firebase(self):
        """Sync session state to Firebase every 10 seconds"""
        if not self.is_active:
            return

        logger.debug("Syncing session to Firebase")

        sync_data = {
            'lastHeartbeat': datetime.now().isoformat(),
            'timeUsed': self.time_used
        }

        # Update session
        session_result = self.firebase.db_update(
            f'sessions/{self.session_id}',
            sync_data
        )

        # Update user's remaining time
        user_result = self.firebase.db_update(
            f'users/{self.user_id}',
            {'remainingTime': self.remaining_time}
        )

        if session_result.get('success') and user_result.get('success'):
            logger.debug("Sync successful")

            # Reset failure counter
            if self.consecutive_sync_failures > 0:
                self.consecutive_sync_failures = 0
                self.is_online = True
                self.sync_restored.emit()
                logger.info("Connection restored")
        else:
            # Handle sync failure
            self.consecutive_sync_failures += 1
            logger.error(f"Sync failed ({self.consecutive_sync_failures} times)")

            if self.consecutive_sync_failures >= 3:
                self.is_online = False
                self.sync_failed.emit("Connection lost")

            # Queue update for retry
            self.sync_queue.append({
                'timestamp': time.time(),
                'data': sync_data
            })

    def _final_sync(self, reason: str):
        """Final sync when ending session"""
        logger.info("Final session sync")

        # Update session
        session_data = {
            'endTime': datetime.now().isoformat(),
            'isActive': False,
            'status': 'completed' if reason == 'user' else reason,
            'timeUsed': self.time_used,
            'lastHeartbeat': datetime.now().isoformat()
        }

        self.firebase.db_update(f'sessions/{self.session_id}', session_data)

        # Update user
        self.firebase.db_update(f'users/{self.user_id}', {
            'remainingTime': max(0, self.remaining_time),
            'activeSessionId': None
        })

    def get_remaining_time(self) -> int:
        """Get current remaining time in seconds"""
        return self.remaining_time

    def get_time_used(self) -> int:
        """Get time used in current session"""
        return self.time_used

    def is_session_active(self) -> bool:
        """Check if session is active"""
        return self.is_active

File: sionyx-kiosk-wpf/src/SionyxKiosk/Services/SessionService.cs
using System.Text.Json;
using System.Windows.Threading;
using SionyxKiosk.Infrastructure;

namespace SionyxKiosk.Services;

/// <summary>
/// Manages user sessions with Firebase sync, countdown, warnings,
/// print monitoring, operating hours, and session lifecycle.
/// Replaces both session_service.py and session_manager.py.
/// </summary>
public class SessionService : BaseService, IDisposable
{
    protected override string ServiceName => "SessionService";

    private readonly ComputerService _computerService;
    private string _userId;
    private readonly string _orgId;

    // Sub-services
    public OperatingHoursService OperatingHours { get; }

    // Session state
    public string? SessionId { get; private set; }
    public bool IsActive { get; private set; }
    public int RemainingTime { get; private set; }
    public int TimeUsed { get; private set; }
    public DateTime? StartTime { get; private set; }

    // Wall-clock anchor for drift-free countdown
    private int _initialRemainingTime;

    // Sync state
    private int _consecutiveSyncFailures;
    public bool IsOnline { get; private set; } = true;

    // Timers (WPF DispatcherTimer for UI thread safety)
    private readonly DispatcherTimer _countdownTimer;
    private readonly DispatcherTimer _syncTimer;

    // Warning flags
    private bool _warned5Min;
    private bool _warned1Min;

    // Events (replace PyQt signals)
    public event Action? SessionStarted;
    public event Action<int>? TimeUpdated;           // remaining seconds
    public event Action<string>? SessionEnded;        // reason: expired, user, error, hours
    public event Action? Warning5Min;
    public event Action? Warning1Min;
    public event Action<string>? SyncFailed;          // error message
    public event Action? SyncRestored;
    public event Action<int>? OperatingHoursWarning;  // minutes until closing
    public event Action<string>? OperatingHoursEnded;  // grace behavior

    public SessionService(FirebaseClient firebase, string userId, string orgId)
        : base(firebase)
    {
        _userId = userId;
        _orgId = orgId;
        _computerService = new ComputerService(firebase);

        // Operating hours
        OperatingHours = new OperatingHoursService(firebase);
        OperatingHours.HoursEndingSoon += OnHoursEndingSoon;
        OperatingHours.HoursEnded += OnHoursEnded;

        // Tick at 250ms so the display never visually skips seconds,
        // even when the UI thread is briefly busy with animations/rendering.
        _countdownTimer = new DispatcherTimer(DispatcherPriority.Render)
        {
            Interval = TimeSpan.FromMilliseconds(250)
        };
        _countdownTimer.Tick += (_, _) => OnCountdownTick();

        _syncTimer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(60) };
        _syncTimer.Tick += (_, _) => _ = SyncToFirebaseAsync();

        Logger.Information("Session service initialized for user: {UserId}", userId);
    }

    /// <summary>Update userId for a new login session (singleton reuse).</summary>
    public void Reinitialize(string userId)
    {
        _userId = userId;
        Logger.Information("Session service re-initialized for user: {UserId}", userId);
    }

    /// <summary>Start a new session with the user's remaining time.</summary>
    public async Task<ServiceResult> StartSessionAsync(int initialRemainingTime)
    {
        if (IsActive) return Error("Session already active");

        // Check time expiration
        var expired = await CheckTimeExpirationAsync();
        if (expired) return Error("הזמן שלך פג תוקף. אנא רכוש חבילה חדשה.");

        if (initialRemainingTime <= 0) return Error("No time remaining");

        // Clean up previous processes
        await Task.Run(() =>
        {
            var cleanup = new ProcessCleanupService();
            cleanup.CleanupUserProcesses();
        });

        // Mark session active in Firebase
        var now = DateTime.Now.ToString("o");
        var result = await Firebase.DbUpdateAsync($"users/{_userId}", new
        {
            isSessionActive = true,
            sessionStartTime = now,
            updatedAt = now,
        });

        if (!result.Success) return Error("Failed to start session");

        // Initialize local state
        SessionId = Guid.NewGuid().ToString("N");
        _initialRemainingTime = initialRemainingTime;
        RemainingTime = initialRemainingTime;
        TimeUsed = 0;
        StartTime = DateTime.UtcNow;
        IsActive = true;
        _warned5Min = false;
        _warned1Min = false;

        // Start timers
        _countdownTimer.Start();
        _syncTimer.Start();

        // Start operating hours
        OperatingHours.StartMonitoring();

        Logger.Information("Session started (remaining: {Time}s)", initialRemainingTime);
        SessionStarted?.Invoke();
        return Success(new { SessionId });
    }

    /// <summary>End the current session.</summary>
    public async Task<ServiceResult> EndSessionAsync(string reason = "user")
    {
        if (!IsActive) return Error("No active session");

        Logger.Information("Ending session: {Reason}", reason);

        // Stop timers
        _countdownTimer.Stop();
        _syncTimer.Stop();

        // Stop monitoring
        OperatingHours.StopMonitoring();

        // Final sync
        await FinalSyncAsync(reason);

        // Browser cleanup (async in background)
        await Task.Run(() =>
        {
            var browserCleanup = new BrowserCleanupService();
            browserCleanup.CleanupWithBrowserClose();
        });

        IsActive = false;
        SessionEnded?.Invoke(reason);

        Logger.Information("Session ended (used: {TimeUsed}s)", TimeUsed);
        return Success(new { TimeUsed, RemainingTime });
    }

    public void Dispose()
    {
        _countdownTimer.Stop();
        _syncTimer.Stop();
        OperatingHours.Dispose();
        GC.SuppressFinalize(this);
    }

    // ==================== PRIVATE HELPERS ====================

    private void OnCountdownTick()
    {
        if (!IsActive || StartTime == null) return;

        // Derive from wall-clock — immune to DispatcherTimer drift
        var elapsed = (int)(DateTime.UtcNow - StartTime.Value).TotalSeconds;
        var newRemaining = Math.Max(0, _initialRemainingTime - elapsed);

        // Only push UI updates when the displayed second actually changes
        if (newRemaining == RemainingTime && elapsed == TimeUsed) return;

        TimeUsed = elapsed;
        RemainingTime = newRemaining;
        TimeUpdated?.Invoke(RemainingTime);

        if (RemainingTime <= 300 && !_warned5Min)
        {
            _warned5Min = true;
            Warning5Min?.Invoke();
        }
        if (RemainingTime <= 60 && !_warned1Min)
        {
            _warned1Min = true;
            Warning1Min?.Invoke();
        }
        if (RemainingTime <= 0)
        {
            _ = EndSessionAsync("expired");
        }
    }

    private async Task SyncToFirebaseAsync()
    {
        if (!IsActive) return;

        var result = await Firebase.DbUpdateAsync($"users/{_userId}", new
        {
            remainingTime = RemainingTime,
            updatedAt = DateTime.Now.ToString("o"),
        });

        if (result.Success)
        {
            if (_consecutiveSyncFailures > 0)
            {
                _consecutiveSyncFailures = 0;
                IsOnline = true;
                SyncRestored?.Invoke();
            }
        }
        else
        {
            _consecutiveSyncFailures++;
            Logger.Error("Sync failed ({Count} consecutive)", _consecutiveSyncFailures);
            if (_consecutiveSyncFailures >= 3)
            {
                IsOnline = false;
                SyncFailed?.Invoke("Connection lost");
            }
        }
    }

    private async Task FinalSyncAsync(string reason)
    {
        await Firebase.DbUpdateAsync($"users/{_userId}", new Dictionary<string, object?>
        {
            ["remainingTime"] = Math.Max(0, RemainingTime),
            ["isSessionActive"] = false,
            ["sessionStartTime"] = null,
            ["updatedAt"] = DateTime.Now.ToString("o"),
        });
    }

    private async Task<bool> CheckTimeExpirationAsync()
    {
        try
        {
            var result = await Firebase.DbGetAsync($"users/{_userId}");
            if (!result.Success || result.Data is not JsonElement data || data.ValueKind == JsonValueKind.Null)
                return false;

            var expiresAtStr = data.TryGetProperty("timeExpiresAt", out var te) ? te.GetString() : null;
            if (string.IsNullOrEmpty(expiresAtStr)) return false;
            if (!DateTime.TryParse(expiresAtStr, out var expiresAt)) return false;

            if (DateTime.Now <= expiresAt) return false;

            // Time expired - reset to 0
            await Firebase.DbUpdateAsync($"users/{_userId}", new Dictionary<string, object?>
            {
                ["remainingTime"] = 0,
                ["timeExpiresAt"] = null,
                ["updatedAt"] = DateTime.Now.ToString("o"),
            });
            return true;
        }
        catch (Exception ex)
        {
            Logger.Error(ex, "Error checking time expiration");
            return false;
        }
    }

    private void OnHoursEndingSoon(int minutes)
    {
        Logger.Warning("Operating hours ending in {Minutes} minutes", minutes);
        OperatingHoursWarning?.Invoke(minutes);
    }

    private void OnHoursEnded(string graceBehavior)
    {
        Logger.Warning("Operating hours ended, behavior: {Behavior}", graceBehavior);
        OperatingHoursEnded?.Invoke(graceBehavior);

        var reason = graceBehavior == "force" ? "hours_force" : "hours";
        _ = EndSessionAsync(reason);
    }
}
